### ISSUE TITLE
Bring BS440 into a Docker container

### ISSUE THREAD

**Issue by Boldfor** [keptenkurk/BS440#106]
# Situation
One comfortable way to run a smarthome server and all required software is via docker. Depending on your smarthome system (Home Assistant, openHAB, etc) a pre-defined docker-compose.yml file helps to set up a pre-configured system incl. all software (influxdb, mosquitto, etc) in a very short amount of time (e.g. see [here](https://community.openhab.org/t/install-maintain-openhab-on-raspi-in-docker-together-with-mosquitto-duplicati-influxdb-samba-etc-step-by-step-instruction/141689)). In such a setting, the BS440-script is the only thing that has to be installed manually.

# Idea
Containerize the BS440 script incl. all required installations, so that the BS440-script can be installed among all other required software in just one docker stack.

# Required steps (?!?)
1. Creating a Dockerfile
2. Specify the base image to use
3. Copy the files from the project into the container
4. Run to install any dependencies
5. Specify the command that should be run when the container is started
6. Build the docker image
7. Push the image to a container registry (e.g. Docker Hub)
8. Run the image
9. ...?

# Remarks
Achieving this would be pretty neat (maybe not only for my setup?). In case someone can give hints / can contribute, let me know...

**Comment by Boldfor** [comment]
Hi @DjZU, @creatvty, @keptenkurk, @poelzi & @udo1toni, I'm writing to you since you've been active in this BS440 repo in the past. Quick question for you.

I used this project two years back to learn python (hence my PRs, which focused mainly on small fixes and code documentations). Now my current learning project is to create docker images.

Since this project appears to be rather low in complexity, and no docker image is available for this project yet (_and_ it's the only component in my home automation stack that's not on docker), I'll be trying to build a docker image out of this nice little script.

Assuming that this would lead anywhere (there's also a fair chance that I'll fail miserably ;) ), would you be willing to test with what I'll hopefully come up with?

Alternatively, if you think that what I'll be trying to do is a bad idea (e.g. there's a low chance for it to work without re-writing large parts of the code) let me know. It'll then save me lots of time.

**Comment by DjZU** [comment]
Hi @Boldfor,

Unlike the author @keptenkurk who uses this BS440 script intermittently on a Raspberry Pi (please do correct me if I'm wrong), I use it on a HTPC/Webserver running 24/7, and faced a number of issues specific to that context, related to the interaction of the system with the Bluetooth adapter, and the pygatt library. I solved those tampering with the pygatt library in a Python virtual environment, tampering with the BS440 code, creating a systemd service and a cron job, as you are aware of that I described in #95 and #110. I am now happy to enjoy a stable version. 

I do use Docker even though I am not a big fan of it, let alone an expert... I am glad to see activity on this repo though, and I will happily give a try to your Docker port to support your effort in my free time (which is obviously limited). It still looks challenging when you read things like this: https://forums.docker.com/t/list-hci-interfaces-with-bluez-hciconfig-fails/136882

Have fun with this!

**Comment by Boldfor** [comment]
Thanks @DjZU, I guess you are right. I started working on the image some days back, and the first major roadblock that I'm hitting is (surprise) in the area of Bluetooth. :D

Let's see... I'll keep you posted.

**Comment by keptenkurk** [comment]
@DjZU You are right, I still use the script on a daily basis for the last 10 years now. I restart is sometime before my shceduled wakeup time. And on the occasion it does not work: no sweat, the data will be pulled from the scale on the next run as it will store 30 weighings.
Cool to see that there is still interest in the code that i wrote 10 years ago!

**Comment by Boldfor** [comment]
@keptenkurk, since you have probably more experience with BT and your script (and maybe also some experience with docker): Do you think it's possible what I'm trying to do, specifically on the BT part?

**Comment by keptenkurk** [comment]
My BT experience dates back to 2015 i'm afraid.

**Comment by Boldfor** [comment]
@keptenkurk , I'm making progress with a Docker-version of the script. The code, however, looks *substantially* different than before (more on that later).

Would you want me to issue a PR (accepting that the original non-Docker-version would be gone in your repo), or shall I just fork your repo and have the new Docker-version in my repo, crediting your work?

**Comment by Boldfor** [comment]
Update: First Version brought online (non-polished with only one plugin for mqtt).

https://github.com/Boldfor/BS440-DOCKER
